File: frontend-vue/src/services/apiService.ts
import axios, { AxiosResponse, AxiosError } from 'axios';
import { 
  ApiResponse, 
  ImageUploadResponse,
  ImageGenerationResponse
} from '@/types';

import { CONFIG } from '@/utils/config';

const API_BASE_URL = CONFIG.baseURL || '/api';
const IMAGE_BASE_URL = CONFIG.imageBaseURL ? 
  (CONFIG.imageBaseURL.endsWith('/') ? CONFIG.imageBaseURL : `${CONFIG.imageBaseURL}/`) : 
  '/';

// 创建axios实例
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  },
  timeout: 60000, // 60秒超时
});

// 请求拦截器
apiClient.interceptors.request.use(
  config => {
    // 可以在这里添加认证token等
    return config;
  },
  (error: AxiosError) => {
    return Promise.reject(error);
  }
);

// 响应拦截器
apiClient.interceptors.response.use(
  (response: AxiosResponse) => {
    return response;
  },
  (error: AxiosError) => {
    const { response } = error;
    let message = '未知错误';
    
    if (response) {
      const { status, data } = response;
      message = (data as any).message || `请求错误: ${status}`;
    } else if (error.request) {
      message = '服务器无响应';
    } else {
      message = error.message || '请求出错';
    }
    
    // 输出错误到控制台，而不是使用ElMessage
    console.error(`API错误: ${message}`);
    return Promise.reject(error);
  }
);

const apiService = {
  /**
   * 获取完整的图片URL
   * @param {string} imagePath - 图片路径
   * @returns {string} 完整的图片URL
   */
  getFullImageUrl(imagePath: string): string {
    if (!imagePath) return '';
    if (imagePath.startsWith('http') || imagePath.startsWith('data:')) {
      return imagePath;
    }
    
    // 确保路径正确
    const path = imagePath.startsWith('/') ? imagePath.slice(1) : imagePath;
    
    // 如果路径中没有static，添加static路径
    if (!path.startsWith('static/') && !path.startsWith('/static/')) {
      return `${IMAGE_BASE_URL}static/${path}`;
    }
    
    return `${IMAGE_BASE_URL}${path}`;
  },

  /**
   * 上传图片
   * @param {File} file - 要上传的文件
   * @returns {Promise} 包含上传结果的Promise
   */
  async uploadImage(file: File): Promise<ImageUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);
    
    const response = await apiClient.post<ImageUploadResponse>(`/upload`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    
    return response.data;
  },

  /**
   * 发送文本生成图片
   * @param {string} text - 用户输入的文本
   * @param {string|null} imageId - 原始图片ID，用于编辑模式
   * @param {boolean} continuousEdit - 是否启用连续编辑模式
   * @returns {Promise} 包含生成结果的Promise
   */
  async generateImage(
    text: string, 
    imageId: string | number | null = null, 
    continuousEdit: boolean = false
  ): Promise<ApiResponse<ImageGenerationResponse>> {
    const payload = {
      text,
      imageId,
      continuousEdit,
    };
    
    const response = await apiClient.post<ApiResponse<ImageGenerationResponse>>('/generate', payload);
    return response.data;
  },

  /**
   * 文本生成图片
   * @param prompt 提示词
   * @returns 生成结果
   */
  async generateFromText(prompt: string): Promise<AxiosResponse<ImageGenerationResponse>> {
    return apiClient.post<ImageGenerationResponse>('/generate_from_text', {
      prompt: prompt
    });
  },

  /**
   * 基于图片生成图片
   * @param file 图片文件
   * @param prompt 提示词
   * @returns 生成结果
   */
  async generateFromImage(file: File, prompt: string): Promise<AxiosResponse<ImageGenerationResponse>> {
    // 记录详细日志
    console.log(`准备发送图片生成请求 - 文件: ${file.name}, 类型: ${file.type}, 大小: ${Math.round(file.size / 1024)}KB`);
    
    const formData = new FormData();
    formData.append('file', file);
    formData.append('prompt', prompt);
    
    // 显示明确的请求信息
    console.log('请求地址:', `${apiClient.defaults.baseURL}/generate`);
    
    return apiClient.post<ImageGenerationResponse>('/generate', formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      },
      // 增加超时时间
      timeout: 90000 // 90秒
    });
  },

  /**
   * 下载图片
   * @param {string} imageUrl - 图片URL
   * @param {string} filename - 下载的文件名
   */
  downloadImage(imageUrl: string, filename?: string): void {
    const fullUrl = this.getFullImageUrl(imageUrl);
    const link = document.createElement('a');
    link.href = fullUrl;
    link.download = filename || 'generated-image.png';
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
  },
};

export default apiService; 

File: ai_image_generator/lib/services/api_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:http_parser/http_parser.dart';
import 'package:flutter/foundation.dart';
import '../models/api_response.dart';
import '../utils/config.dart';

class ApiService {
  static final ApiService _instance = ApiService._internal();
  late final Dio _dio;
  final Config _config = Config();

  factory ApiService() {
    return _instance;
  }

  ApiService._internal() {
    _dio = Dio(BaseOptions(
      baseUrl: _config.baseURL,
      connectTimeout: const Duration(seconds: 60),
      receiveTimeout: const Duration(seconds: 60),
      headers: {
        'Accept': 'application/json',
      },
    ));

    // Add interceptors for logging if in debug mode
    if (kDebugMode) {
      _dio.interceptors.add(LogInterceptor(
        request: true,
        requestHeader: true,
        requestBody: true,
        responseHeader: true,
        responseBody: true,
        error: true,
      ));
    }
  }

  // Get full image URL
  String getFullImageUrl(String imagePath) {
    return _config.getFullImageUrl(imagePath);
  }

  // Generate image from text
  Future<ImageGenerationResponse> generateFromText(String prompt) async {
    try {
      final response = await _dio.post('/generate_from_text', data: {
        'prompt': prompt,
      });

      return ImageGenerationResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('Error generating image from text: ${e.message}');
      return ImageGenerationResponse(
        success: false,
        message: e.message,
      );
    }
  }

  // Generate image from existing image
  Future<ImageGenerationResponse> generateFromImage(
      File image, String prompt) async {
    try {
      // 确定文件的MIME类型
      String mimeType = 'image/jpeg'; // 默认类型
      String filename = image.path.split('/').last;

      // 根据文件扩展名设置对应的MIME类型
      if (filename.toLowerCase().endsWith('.png')) {
        mimeType = 'image/png';
      } else if (filename.toLowerCase().endsWith('.jpg') ||
          filename.toLowerCase().endsWith('.jpeg')) {
        mimeType = 'image/jpeg';
      } else if (filename.toLowerCase().endsWith('.webp')) {
        mimeType = 'image/webp';
      }

      // 如果文件没有扩展名或不是支持的格式，添加.jpg扩展名
      if (!filename.toLowerCase().endsWith('.jpg') &&
          !filename.toLowerCase().endsWith('.jpeg') &&
          !filename.toLowerCase().endsWith('.png') &&
          !filename.toLowerCase().endsWith('.webp')) {
        filename = "$filename.jpg";
      }

      print('上传图片: $filename, MIME类型: $mimeType');

      FormData formData = FormData.fromMap({
        'file': await MultipartFile.fromFile(
          image.path,
          filename: filename,
          contentType: MediaType.parse(mimeType),
        ),
        'prompt': prompt,
      });

      final response = await _dio.post(
        '/generate',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
          sendTimeout: const Duration(seconds: 90),
        ),
      );

      return ImageGenerationResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('Error generating image from image: ${e.message}');
      return ImageGenerationResponse(
        success: false,
        message: e.message,
      );
    }
  }

  // Upload image
  Future<ImageUploadResponse> uploadImage(File file) async {
    try {
      FormData formData = FormData.fromMap({
        'file': await MultipartFile.fromFile(
          file.path,
          filename: file.path.split('/').last,
        ),
      });

      final response = await _dio.post(
        '/upload',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        ),
      );

      return ImageUploadResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('Error uploading image: ${e.message}');
      return ImageUploadResponse(
        success: false,
        message: e.message,
      );
    }
  }
}
